bam_markdup.c

```diff
@@ -943,118 +1086,227 @@ static int duplicate_chain_check(md_param_t *param, khash_t(duplicates) *dup_has
+                                    ret = -1;
+                                    goto fail;
+                                }
 
-                    if (is_cur_dup) {
-                        // the current is the optical duplicate
-                        if (!is_cur_opt) { // only change if not already an optical duplicate
-                            if (optical_retag(param, dup_hash, current->b, current_paired, optical_single, optical_pair)) {
-                                ret = -1;
-                                break;
+                                chk->opt = 1;
                             }
-                        }
-                    } else {
-                        if (!is_ori_opt) {
-                            if (optical_retag(param, dup_hash, ori->b, ori_paired, optical_single, optical_pair)) {
-                                ret = -1;
-                                break;
+                        } else {
+                            if (!current->opt) {
+                                if (optical_retag(param, dup_hash, current->b, current_paired, optical_single, optical_pair)) {
+                                    ret = -1;
+                                    goto fail;
+                                }
+
+                                current->opt = 1;
                             }
                         }
                     }
                 }
             }
+
+            count++;
         }
 
-        current = current->duplicate;
+        curr++;
     }
 
+ fail:
     return ret;
 }
 
+
+/* Where there is more than one duplicate go down the list and check for optical duplicates and change
+   do tags (where used) to point to original (non-duplicate) read. */
+static int find_duplicate_chains(md_param_t *param, klist_t(read_queue) *read_buffer, khash_t(duplicates) *dup_hash, check_list_t *dup_list,
+                                const hts_pos_t prev_coord, const int32_t prev_tid, long *warn, long *optical_single,
+                                long *optical_pair, const int check_range) {
+    int ret = 0;
+    kliter_t(read_queue) *rq;
+    long tmp_count = 0;
+
+    rq = kl_begin(read_buffer);
+
+    while (rq != kl_end(read_buffer)) {
+        read_queue_t *in_read = &kl_val(rq);
+
+        if (check_range) {
+            /* Just check against the moving window of reads based on coordinates and max read length. */
+            if (in_read->pos + param->max_length > prev_coord && in_read->b->core.tid == prev_tid && (prev_tid != -1 || prev_coord != -1)) {
+                break;
+            }
+        } else {
+            // this is the last set of results and the end entry will be blank
+            if (!bam_get_qname(in_read->b)) {
+                break;
+            }
+        }
+
+        if (!(in_read->b->core.flag & BAM_FDUP) && in_read->duplicate) { // is the head of a duplicate chain
```

Comment:
On a small test file it still seems to spend longer than I expected, with about 50% of time at this line of code surprisingly.

It got about 20% quicker just by swapping the order of conditionals, ie:

```
if (in_read->duplicate && !(in_read->b->core.flag & BAM_FDUP))
```

I'm not sure if my test data was abnormal in some way though.

Reply:
Ignore this.  While still more efficient the other way around, I think it *shouldn't* be spending so long in this function.  It's a symptom of my input file having already marked for duplicates and not using the "-c" option to clear those flags.

I don't know yet why that makes such a difference though.